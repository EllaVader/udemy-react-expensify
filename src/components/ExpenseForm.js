import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';import 'react-dates/lib/css/_datepicker.css';


//const date = new Date();
const now = moment();
console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
  // use local component state to track changes to user inputs.
  // only when the user submits the form will we send them via redux
  state = {
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    calendarFocused: false
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({note}));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    //only allow for numbers followed by an optional . followed by 2 numbers
    if (amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({amount}));
    }
  }

  onDateChanged = (createdAt) => {
    this.setState(() => ({createdAt}));
  }

  onFocusChanged = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}))
  }

  render(){
    return (
      <div>
        <form>
          <input type="text" placeholder="Description" autoFocus value={this.state.description} onChange={this.onDescriptionChange}/>
          <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange}/>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChanged}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChanged}
            numberOfMonths={1}
            isOutsideRange={() => false }
          />
          <textarea value={this.state.note} onChange={this.onNoteChange} placeholder="Add a note for your expense (optional)"></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}