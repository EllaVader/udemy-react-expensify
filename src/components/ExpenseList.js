import React from 'react';
import { connect } from 'react-redux';

// our regular unconnected component
// pass in the props from the state that we connected to in the connect method below.
const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    { props.filters.text}
    { props.expenses.length }
  </div>
);

// map the state object to component props.
const mapStateToProps = (state) => {
  // return an object that contains the things we want to access from the store.
  return {
    expenses: state.expenses,
    filters: state.filters
  }
};

// create a higher order component that connects our component to the store.
// returns a function, pass the component into the function
// in the connect fn the argument is info to what we want to connect
// lets us determine what info from the store we want our component to be able to access
export default connect(mapStateToProps)(ExpenseList);  