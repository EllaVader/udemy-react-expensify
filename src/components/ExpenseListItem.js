import React from 'react';

// one way to do it 
// const ExpenseListItem = (props) => (
//   <div>
//     <h3>{props.expense.description}</h3>
//     <p>{props.expense.amount} - {props.expense.createdAt}</p>
//   </div>
// );

// another way to do it - we spread the expenses in the ExpenseList component
// and we are destructuring them here in the argument list.
const ExpenseListItem = ({description, amount, createdAt}) => (
  <div>
    <h3>{description}</h3>
    <p>{amount} - {createdAt}</p>
  </div>
);


export default ExpenseListItem;