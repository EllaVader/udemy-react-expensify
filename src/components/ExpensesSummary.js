import React from 'react';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import { connect } from 'react-redux';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div>
      <h2>
        Viewing {expenseCount} {expenseWord} totaling {formattedExpensesTotal}
      </h2>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  const expensesTotal = selectExpensesTotal(visibleExpenses);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal
  }
}

export default connect(mapStateToProps)(ExpensesSummary); ;
