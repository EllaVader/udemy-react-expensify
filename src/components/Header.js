import React from 'react';
import { NavLink } from 'react-router-dom';

// set up our individual routes via JSX
// must set root page to exact=true so it won't pattern match when navigating to other pages that contain "/"
const Header = () => (
  <header>
    <h1>Expensify</h1>
    <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
    <NavLink to="/create" activeClassName="is-active">Add Expense</NavLink>
    <NavLink to="/edit" activeClassName="is-active">Edit Expense</NavLink>
    <NavLink to="/help" activeClassName="is-active">Help</NavLink>
  </header>
)

export default Header;