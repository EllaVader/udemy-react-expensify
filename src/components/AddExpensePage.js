import React from 'react';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';
import { connect } from 'react-redux';

// we will connect this component to the store so it can add a new expense.

const AddExpensePage = (props) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm 
      onSubmit={(expense) => {
        props.dispatch(addExpense(expense));
        // using browser routing go to dashboard page
        props.history.push('/');
      }}
      />
  </div>
);

export default connect()(AddExpensePage);