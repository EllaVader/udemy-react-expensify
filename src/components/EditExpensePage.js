import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';


const EditExpensePage = (props) => {
  // props.match.params -- this is a special property on the Route object called "match" when we user Route.
  // id was passed in as one of the parameters for this page Route path="/edit/:id"  we can access
  // it via the props.match
  return (
    <div>
      <ExpenseForm
        expense={props.expense} 
        onSubmit={(expense) => {
          props.dispatch(editExpense(props.expense.id, expense));
          // using browser routing go to dashboard page. we have access to this history prop and it has a push method on it to go to a different page.
          props.history.push('/');
          console.log('updated', expense);
        }}
      />
      <button onClick={() => {
        props.dispatch(removeExpense({ id: props.expense.id }));
        props.history.push('/');
      }}>Remove</button>
  </div>
  );
}

const mapStateToProps = (state, props) => {
  // return an object that contains the things we want to access from the store
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
};

export default connect(mapStateToProps)(EditExpensePage);
