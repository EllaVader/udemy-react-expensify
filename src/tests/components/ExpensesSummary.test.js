import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render ExpensesSummary with 1 expense', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={1234} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary more than one expense', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={2} expensesTotal={1234876} />);
  expect(wrapper).toMatchSnapshot();
});