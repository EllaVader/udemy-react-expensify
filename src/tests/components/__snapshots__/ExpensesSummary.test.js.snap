// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary more than one expense 1`] = `
<div>
  <h2>
    Viewing 
    2
     
    expenses
     totaling 
    $12,348.76
  </h2>
</div>
`;

exports[`should render ExpensesSummary with 1 expense 1`] = `
<div>
  <h2>
    Viewing 
    1
     
    expense
     totaling 
    $12.34
  </h2>
</div>
`;
